### PR TITLE
Add default paragraph graphic factory

### LINKED ISSUE #151: Lists do not show numbers or bullets
When marking text rows as a list we are not seeing any graphics representing the items.

### LINKED ISSUE #152: Numbers and bullets alignment and layout issues.
As can be seen in the screenshot below:
- The bullets and numbers do not align nicely with the actual text.
- It seems that numbers sometimes become truncated, as you can see it displays "..." instead of the number. Haven't seen this if I increase the indentation.

These issues become more noticeable if you increate the font size a bit from the default one.

image

Below is a sample from the osx TextEdit app.

image

### PR DESCRIPTION
**Pull request by jperedadnr**
Fixes #151 
Fixes #152

### CODE REVIEW

File: src/main/java/com/gluonhq/richtextarea/model/ParagraphDecoration.java [hunk 1]
```diff
@@ -183,11 +183,13 @@ public Builder leftInset(double leftInset) {
 
         public Builder indentationLevel(int indentationLevel) {
             this.indentationLevel = indentationLevel;
+            this.graphicType = indentationLevel == 0 ? GraphicType.NONE : graphicType;
             return this;
         }
 
         public Builder graphicType(GraphicType graphicType) {
             this.graphicType = graphicType;
+            this.indentationLevel = Math.max(graphicType != GraphicType.NONE ? 1 : 0, indentationLevel);
```
Comment (abhinayagarwal) [inline src/main/java/com/gluonhq/richtextarea/model/ParagraphDecoration.java:192]: I would reverse the condition to `graphicType == GraphicType.NONE ? 0 : 1`

File: src/main/java/com/gluonhq/richtextarea/ParagraphTile.java [hunk 2]
```diff
@@ -145,40 +141,79 @@ TextFlow getTextFlow() {
...
+    private void updateBox(BiFunction<Integer, ParagraphDecoration.GraphicType, Node> graphicFactory) {
+        graphicBox.getChildren().clear();
+
+        ParagraphDecoration decoration = paragraph.getDecoration();
+        int indentationLevel = decoration.getIndentationLevel();
+        Node graphicNode = null;
+        if (graphicFactory != null) {
+            graphicNode = graphicFactory.apply(indentationLevel, decoration.getGraphicType());
+        }
+        double spanPrefWidth = Math.max((indentationLevel - (graphicNode != null ? 1 : 0)) * INDENT_PADDING, 0d);
```
Comment (abhinayagarwal) [inline src/main/java/com/gluonhq/richtextarea/ParagraphTile.java:153]: I would reverse the condition to `graphicNode == null ? 0 : 1`

**Comment by abhinayagarwal** [comment]
Some more findings:

* It is not possible to remove a bullet by toggling off the button
* Pressing backspace on the start of the bullet should only remove the bullet. Currently, it moves to previous paragraph.

File: src/main/java/com/gluonhq/richtextarea/model/ParagraphDecoration.java (hunk 1 above)
Comment (jperedadnr) [reply src/main/java/com/gluonhq/richtextarea/model/ParagraphDecoration.java:192]: done

File: src/main/java/com/gluonhq/richtextarea/ParagraphTile.java (hunk 2 above)
Comment (jperedadnr) [reply src/main/java/com/gluonhq/richtextarea/ParagraphTile.java:153]: done

**Comment by jperedadnr** [comment]
Right, I can reproduce both. 
First issue should be easy to fix.
The backspace one might be harder.

File: pom.xml [hunk 3]
```diff
@@ -94,6 +94,9 @@
                 <version>${javafx.plugin.version}</version>
                 <configuration>
                     <mainClass>${main.class}</mainClass>
+                    <options>
+                        <option>--add-exports=javafx.graphics/com.sun.javafx.tk=ALL-UNNAMED</option>
```
Comment (dlemmermann) [inline pom.xml:98]: Could this please be turned into a motivation to move some of that API into the public area?

**Review by dlemmermann — APPROVED** [review]

File: pom.xml (hunk 3 above)
Comment (jperedadnr) [reply pom.xml:98]: Right, but you know how it goes, making public exiting private API is challenging to say the least...
For instance, demand for public `FontMetrics` dates back to 2010...  https://bugs.openjdk.java.net/browse/JDK-8090775
The last comment about using the old com/sun/javafx/scene/control/skin/Utils.java gave me an idea, and I've been able to remove the private API...

**Comment by jperedadnr** [comment]
@abhinayagarwal I've addressed your comments, PR is ready for review again.

**Review by abhinayagarwal — APPROVED** [review]
The bullet numbers do not follow the font family/size of the paragraph it is applied to. Although, the fix may be for a separate PR.

**Pull request merged**